process: refactor emitWarning

BridgeAR opened the pull request:
Currently the code does a couple of checks to often even though they
are not necessary.

Also remove a double whitespace in a error comment.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

apapirovski commented on lib/internal/process/warning.js:
```diff
@@ -116,24 +116,23 @@ function setupProcessWarnings() {
       ctor = type;
       code = undefined;
       type = 'Warning';
+    } else if (type !== undefined && typeof type !== 'string') {
```
I don't think this is the same. If `type` is an object but `type.type` is truthy and not a `string` then this won't throw anymore.

BridgeAR replied:
Addressed.

apapirovski commented on lib/internal/process/warning.js:
```diff
@@ -116,24 +116,23 @@ function setupProcessWarnings() {
-      throw new ERR_INVALID_ARG_TYPE('type', 'string', type);
-    if (warning === undefined || typeof warning === 'string') {
+    }
+    if (typeof warning === 'string') {
```
This removes the possibility of an `undefined` warning that has the other properties specified.

apapirovski replied:
While not documented, as far as I can tell this was requested in https://github.com/nodejs/node/pull/4782

BridgeAR replied:
`process.emitWarning()` produces: `(node:XXXX) Warning`. I am not convinced that this is useful at all. This seems like a bugfix as no args should result in an error. Therefore I would say it is a patch. However, I would also keep this if it makes this semver-major.

Ping @nodejs/tsc please alos give an opinion.

Trott replied:
> `process.emitWarning()` produces: `(node:XXXX) Warning`. I am not convinced that this is useful at all. 

`process.emitWarning(undefined, 'fhqwhgads', 'come-on')` produces the slightly more useful `(node:91812) [come-on] fhqwhgads`. 

I have no idea if anyone uses that pattern or not. Maybe it's time for me to figure out how to use Gzemnid.

In the spirit of the [onboarding-extras.md](https://github.com/nodejs/node/blob/master/doc/onboarding-extras.md):

> be conservative – that is, if a change has the remote chance of breaking something, go for semver-major

It seems like this should be semver-major as it has a "remote change of breaking something".

(If we're not going to call this semver-major, I won't push back, but we really ought to remove the quoted sentence above from the doc since it is clearly not something we abide by in that case.)

BridgeAR commented:
I marked it as semver-major. PTAL.

mcollina commented:
I would expect to see more tests changed as it's semver-major. Can you add some more test that verify the change in behavior?

refack commented:
Can one of the admins verify this patch?

BridgeAR commented:
Rebased due to conflicts. I also added a test case for the changed behavior as requested.

CI https://ci.nodejs.org/job/node-test-pull-request/15379/

Trott commented:
> CI https://ci.nodejs.org/job/node-test-pull-request/15379/

Looks like it's a rare instance of missing the CERTIFY_SAFE checkbox?

TimothyGu commented:
I agree with @apapirovski. This PR was "marketed" as a simple refactoring by its title… and it should stay this way at least for this PR, or we could change the title to make it clear that we are changing the behavior.

BridgeAR commented:
@TimothyGu @apapirovski I changed the description of the PR accordingly.

CI https://ci.nodejs.org/job/node-test-pull-request/15591/

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/15892/ (passed)

Landed in 49681e7414629e0d3fbc3ac5bfbd3fa46725c49c :tada:
